src/include/buffer.h

```diff
@@ -491,12 +486,7 @@ namespace buffer CEPH_BUFFER_API {
       iterator(bl_t *l, unsigned o=0);
       iterator(bl_t *l, unsigned o, list_iter_t ip, unsigned po);
 
-#ifdef BL_BACKWARD_COMPAT
-      void advance(int o);
```

Comment:
my concern is that this could break existing librados clients at compile-time and run-time, which still call `buffer::list::iterator::advance(int)`.

at compile time, we will have:
```
use of deleted function ‘void ceph::list::iterator::advance(int)’
```
at run-time, we will have
```
./test: symbol lookup error: ./test: undefined symbol: _ZN4ceph6buffer3Foo4funcEi
```

Reply:
Yeah, I was afraid about the backward compatibility as too. However, @liewegas pointed out we need to care only between minor releases. If my understanding is correct, then it would be enough just to not backport such changes.

Reply:
we do have compatibility tests which 

1. install librados client of an old version, for example, mimic, using rpm/deb
1. backup the executable of the client to somewhere
1. install librados of current version, i.e., nautilus, using rpm/deb
1. put back the executable
1. run it.

see

- 9b92d87d37e3892b0096cd728b46154aed1e2d86
- f8a8b4f0e41929bb4dde4217ae2bc1e8b02f7b2c
- 77b244fd8e0513be77e701f620c1edf9a9237cc5

for examples for how we broke/fixed the ABI before.

Reply:
Hrm, I thought this was only a problem when we did it in a minor bugfix update.  @jdurgin ?

Reply:
Oh we've made that argument before, but if that's our claim we also probably need to get better about bumping the librados version and shipping multiple versions within each release cycle so that users have a sane upgrade cycle for their clients as well as their servers. :/

Reply:
the reason we don't bump the so version is because these sorts of changes only affect the C++ interface, but librados.so also contains the C library, which libvirt, qemu, and others depend on. 

Since c++ build changes already break the ABI regularly between major versions, I'm fine with doing that. I guess the concern is more around API changes.

Reply:
i managed to break the C++ ABI couple times before. before breaking it again, i think we probably need to take the issues into consideration, otherwise we will need to revert this change later

- http://tracker.ceph.com/issues/21573
- http://tracker.ceph.com/issues/17809
- http://tracker.ceph.com/issues/17809 (this issue resembles the one we could run into if this PR gets merged)

cc @dillaman